File: Rust-SDL2/src/flake.rs
use sdl2::rect::Rect;
use sdl2::render::{Texture, WindowCanvas};
use rand::Rng;


pub struct Flake<'a> {
    texture: &'a Texture<'a>,
    rng: rand::rngs::ThreadRng,
    position: Rect,
    window_width: i32,
    window_height: i32,
    y_pos: f32,
    y_vel: f32,
    pub white: bool,
}

impl<'a> Flake<'a> {
    pub fn new(texture: &'a Texture<'a>, white: bool, window_width: i32, window_height: i32) -> Result<Flake<'a>, String> {
        let y_vel = 300.0;
        let mut rng = rand::thread_rng();
        let mut position = Rect::new(0, 0 as i32, texture.query().width, texture.query().height);
        position.x = rng.gen_range(0..(window_width - position.width() as i32));
        position.y = -(rng.gen_range(0..(window_height * 2)) + position.height() as i32);
        let y_pos = position.y as f32;
        Ok(Flake { texture, white, rng, position, window_width, window_height, y_pos, y_vel })
    }

    pub fn draw(&self, canvas: &mut WindowCanvas) -> Result<(), String> {
        canvas.copy(&self.texture, None, self.position)?;
        Ok(())
    }

    pub fn update(&mut self, delta_time: f32) {
        self.y_pos += self.y_vel * delta_time;
        self.position.y = self.y_pos as i32;
    }

    pub fn left(&self) -> i32 {
        self.position.x
    }

    pub fn right(&self) -> i32 {
        self.position.x + self.position.width() as i32
    }

    pub fn bottom(&self) -> i32 {
        self.position.y + self.position.height() as i32
    }

    pub fn reset(&mut self, full: bool) {
        self.position.x = self.rng.gen_range(0..(self.window_width - self.position.width() as i32));
        if full {
            self.position.y = -(self.rng.gen_range(0..(self.window_height * 2)) + self.position.height() as i32);
        } else {
            self.position.y = -(self.rng.gen_range(0..(self.window_height)) + self.position.height() as i32);
        }
        self.y_pos = self.position.y as f32;
    }
}

File: Go-SDL2/flake.go
package main

import (
	"math/rand"

	"github.com/veandco/go-sdl2/sdl"
)

type flake struct {
	_renderer *sdl.Renderer
	_image    *sdl.Texture
	_rect     sdl.Rect
	_isYellow bool
	_yVel     float64
	_yPos     float64
}

func newFlake(renderer *sdl.Renderer, image *sdl.Texture, isYellow bool) (*flake, error) {
	f := &flake{}
	var err error

	f._renderer = renderer
	f._image = image
	f._isYellow = isYellow
	f._yVel = 300

	if _, _, f._rect.W, f._rect.H, err = f._image.Query(); err != nil {
		return f, err
	}

	f.reset(true)

	return f, err
}

func (f *flake) reset(full bool) {
	if full {
		f._rect.Y = (-f._rect.H - int32(rand.Intn(windowHeight))) * 2
	} else {
		f._rect.Y = -f._rect.H - int32(rand.Intn(windowHeight))
	}
	f._yPos = float64(f._rect.Y)
	f._rect.X = int32(rand.Intn(windowWidth - int(f._rect.W)))
}

func (f *flake) update(dt float64) {
	f._yPos += f._yVel * dt
	f._rect.Y = int32(f._yPos)
}

func (f *flake) left() int32 {
	return f._rect.X
}

func (f *flake) right() int32 {
	return f._rect.X + f._rect.W
}

func (f *flake) bottom() int32 {
	return f._rect.Y + f._rect.H
}

func (f *flake) isYellow() bool {
	return f._isYellow
}

func (f *flake) draw() {
	f._renderer.Copy(f._image, nil, &f._rect)
}
